packages/react-router/src/useBlocker.tsx

```diff
@@ -3,25 +3,62 @@ import { useRouter } from './useRouter'
 import type { BlockerFn } from '@tanstack/history'
 import type { ReactNode } from './route'
 
+type Resolver = {
+  allow: () => void
+  deny: () => void
+}
+
 export function useBlocker(
   blockerFn: BlockerFn,
   condition: boolean | any = true,
-): void {
+) {
   const { history } = useRouter()
 
-  React.useEffect(() => {
-    if (!condition) return
-    return history.block(blockerFn)
+  const [resolver, setResolver] = React.useState<Resolver>({
+    allow: () => {},
+    deny: () => {},
   })
+
+  const createPromise = () =>
+    new Promise<boolean>((resolve) => {
+      setResolver({
+        allow: () => resolve(true),
+        deny: () => resolve(false),
+      })
+    })
+
+  const [promise, setPromise] = React.useState(createPromise)
+
+  React.useEffect(() => {
+    const blockerFnComposed = async () => {
+      const canNavigateSync = blockerFn()
```

Comment:
what happens if `blockerFn` returns a `Promise`?

Reply:
We could await it, but wouldn't that risk blocking the inner promise? Technically, if we fire `.allow()` before we resolve the `blockerFn`, we could end up with the user being stuck in the page.

Adding a `state` property could make `blockerFn` unnecessary as the user would be able to fire an event on state change (effectively giving the user more control over the block).